README.rst
..
  This file is part of AudioLazy, the signal processing Python package.
  Copyright (C) 2012-2016 Danilo de Jesus da Silva Bellini

  AudioLazy is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, version 3 of the License.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.

AudioLazy
=========

.. list-table::
  :stub-columns: 1

  * - Development
    - |travis| |coveralls|
  * - Last release
    - |v| |pyversions| |implementation|
  * - PyPI status
    - |format| |status| |l|

.. |travis| image::
  https://img.shields.io/travis/danilobellini/audiolazy/master.svg
  :target: https://travis-ci.org/danilobellini/audiolazy
  :alt: Travis CI builds

.. |coveralls| image::
  https://img.shields.io/coveralls/danilobellini/audiolazy/master.svg
  :target: https://coveralls.io/r/danilobellini/audiolazy
  :alt: Coveralls coverage report

.. |v| image::
  https://img.shields.io/pypi/v/audiolazy.svg
  :target: https://pypi.python.org/pypi/audiolazy
  :alt: Last stable version (PyPI)

.. |pyversions| image::
  https://img.shields.io/pypi/pyversions/audiolazy.svg
  :target: https://pypi.python.org/pypi/audiolazy
  :alt: Python versions (PyPI)

.. |implementation| image::
  https://img.shields.io/pypi/implementation/audiolazy.svg
  :target: https://pypi.python.org/pypi/audiolazy
  :alt: Python implementations (PyPI)

.. |format| image::
  https://img.shields.io/pypi/format/audiolazy.svg
  :target: https://pypi.python.org/pypi/audiolazy
  :alt: Distribution format (PyPI)

.. |status| image::
  https://img.shields.io/pypi/status/audiolazy.svg
  :target: https://pypi.python.org/pypi/audiolazy
  :alt: Project status (PyPI)

.. |l| image::
  https://img.shields.io/pypi/l/audiolazy.svg
  :target: https://pypi.python.org/pypi/audiolazy
  :alt: License (PyPI)

Real-Time Expressive Digital Signal Processing (DSP) Package for Python!

Laziness and object representation
----------------------------------

There are several tools and packages that let the Python use and
expressiveness look like languages such as MatLab and Octave. However, the
eager evaluation done by most of these tools make it difficult, perhaps
impossible, to use them for real time audio processing. To avoid such
eagerness, one can make the calculations only when data is requested, not
when the path to the data is given. This is the core idea in laziness that
allows:

- Real-time application (you don't need to wait until all data is
  processed to have a result);
- Endless data sequence representation;
- Data-flow representation;
- Task elimination when a reverse task is done: instead of doing something
  to then undo, nothing needs to be done, and no conscious optimization
  need to be done for that.

Another difficulty concerns expressive code creation for audio processing in
blocks through indexes and vectors. Sometimes, that's unavoidable, or at
least such avoidance would limit the power of the system that works with
sequence data.

Block sequences can be found from sample sequences being both objects, where
the latter can be the result of a method or function over the former. The
information needed for such is the block size and where would start the next
block. Although one can think about the last block and the exact index where
it would start, most of the time spent in steps like this one happens to be
an implementation issue that just keep the focus away from the problem being
worked on. To allow a thing like an endless data sequence, there should be
no need to know when something stops.

Probably an engineer would find the use of equations and structures from
electrical engineering theory much cleaner to understand than storing
everything into data arrays, mainly when common operations are done to these
representations. What is the product of the filter with numerator
``[1, 7, 2]`` and denominator ``[1, 0.5, 0.2]`` as its system equation with
the one that has the arrays reversed like ``[2, 7, 1]``? That might be simple,
and the reversed would avoid questions like "what comes first, the zero or the
[minus] two exponent?", but maybe we could get more efficient ourselves if we
had something easier: multiplication could be written once and for all and
with a representation programmers are used to see. This would be even more
expressive if we could get rid from the asymmetry of a method call like
``filt1.multiply_by(filt2)``, since multiplication in this case should be
commutative. The use of natural operators is possible in a language that
allows operator overloading, but for such we need to describe
those equations and structures as objects and object relationships.

The name ``Hz`` can be a number that would allow conversion to a default DSP
internal rad/samples unit, so one can write things like ``freq = 440 * Hz``.
This isn't difficult in probably any language, but can help in expressiveness,
already. If (almost) everything would need data in "samples" or "rad/sample"
units, constants for converting these from "second" and "hertz" would help
with the code expressiveness. A comb filter ``comb.tau(delay=30*s, tau=40*s)``
can represent a comb filter with the given delay and time constant, both in
samples, but with a more clear meaning for the reader than it would have with
an expression like ``[1] + [0] * 239999 + [alpha]``. Would it be needed to
store all those zeros while just using the filter to get a frequency response
plot?

It's possible to avoid some of these problems with well-chosen constants,
duck typing, overloaded operators, functions as first-class citizens, object
oriented together with functional style programming, etc.., resources
that the Python language gives us for free.

What does it do?
----------------

Prioritizing code expressiveness, clarity and simplicity, without precluding
the lazy evaluation, and aiming to be used together with Numpy, Scipy and
Matplotlib as well as default Python structures like lists and generators,
AudioLazy is a package written in pure Python proposing digital audio signal
processing (DSP), featuring:

- A ``Stream`` class for finite and endless signals representation with
  elementwise operators (auto-broadcast with non-iterables) in a common
  Python iterable container accepting heterogeneous data;
- Strongly sample-based representation (Stream class) with easy conversion
  to block representation using the ``Stream.blocks(size, hop)`` method;
- Sample-based interactive processing with ``ControlStream``;
- ``Streamix`` mixer for iterables given their starting time deltas;
- Multi-thread audio I/O integration with PyAudio;
- Linear filtering with Z-transform filters directly as equations (e.g.
  ``filt = 1 / (1 - .3 * z ** -1)``), including linear time variant filters
  (i.e., the ``a`` in ``a * z ** k`` can be a Stream instance), cascade
  filters (behaves as a list of filters), resonators, etc.. Each
  ``LinearFilter`` instance is compiled just in time when called;
- Zeros and poles plots and frequency response plotting integration with
  MatPlotLib;
- Linear Predictive Coding (LPC) directly to ``ZFilter`` instances, from
  which you can find PARCOR coeffs and LSFs;
- Both sample-based (e.g., zero-cross rate, envelope, moving average,
  clipping, unwrapping) and block-based (e.g., window functions, DFT,
  autocorrelation, lag matrix) analysis and processing tools;
- A simple synthesizer (Table lookup, Karplus-Strong) with processing tools
  (Linear ADSR envelope, fade in/out, fixed duration line stream) and basic
  wave data generation (sinusoid, white noise, impulse);
- Biological auditory periphery modeling (ERB and gammatone filter models);
- Multiple implementation organization as ``StrategyDict`` instances:
  callable dictionaries that allows the same name to have several different
  implementations (e.g. ``erb``, ``gammatone``, ``lowpass``, ``resonator``,
  ``lpc``, ``window``);
- Converters among MIDI pitch numbers, strings like "F#4" and frequencies;
- Polynomials, Stream-based functions from itertools, math, cmath, and more!
  Go try yourself! =)

Installing
----------

The package works both on Linux and on Windows. You can find the last stable
version at `PyPI <http://pypi.python.org/pypi/audiolazy>`_ and install it with
the usual Python installing mechanism::

  python setup.py install

If you have pip, you can go directly (use ``-U`` for update or reinstall)::

  pip install audiolazy

for downloading (from PyPI) and installing the package for you, or::

  pip install -U .

To install from a path that has the ``setup.py`` file and the package data
uncompressed previously.

For the *bleeding-edge* version, you can install directly from the github
repository (requires ``git`` for cloning)::

  pip install -U git+git://github.com/danilobellini/audiolazy.git

For older versions, you can install from the PyPI link or directly from the
github repository, based on the repository tags. For example, to install the
version 0.04 (requires ``git`` for cloning)::

  pip install -U git+git://github.com/danilobellini/audiolazy.git@v0.04

The package doesn't have any strong dependency for its core besides the Python
itself (versions 2.7, 3.2 or newer) as well as its standard library, but you
might need:

- PyAudio: needed for playing and recording audio (``AudioIO`` class);
- NumPy: needed for doing some maths, such as finding the LSFs from a filter
  or roots from a polynomial;
- MatPlotLib: needed for all default plotting, like in ``LinearFilter.plot``
  method and several examples;
- SciPy (testing and examples only): used as an oracle for LTI filter testing
  and for the Butterworth filter example;
- Sympy (testing only): used for testing linear filters with time-varying
  matrices of symbolic coeffs where the Stream samples are these matrices;
- tox for testing all at once, or pytest, pytest-cov and pytest-timeout for
  testing in a single environment (testing only): runs test suite and
  shows code coverage status;
- wxPython (example only): used by one example with FM synthesis in an
  interactive GUI;
- Tkinter (example only): needed for the pitch follower based on the
  zero-crossing rate example GUI;
- Music21 (example only): there's one example that gets the Bach chorals from
  that package corpora for synthesizing and playing;
- Sphinx (documentation only): it can create the software documentation in
  several different file formats.

Beside examples and tests, only the filter plotting with ``plot`` and
``zplot`` methods needs MatPlotLib. Also, the routines that needs NumPy up to
now are:

- Root finding with ``zeros`` and ``poles`` properties (filter classes) or
  with ``roots`` property (Poly class);
- Some Linear Predictive Coding (``lpc``) strategies: ``nautocor``,
  ``autocor`` and ``covar``;
- Line Spectral Frequencies ``lsf`` and ``lsf_stable`` functions.

Getting started
---------------

Before all examples below, it's easier to get everything from audiolazy
namespace:

.. code-block:: python

  from audiolazy import *

All modules starts with "lazy\_", but their data is already loaded in the main
namespace. These two lines of code do the same thing:

.. code-block:: python

  from audiolazy.lazy_stream import Stream
  from audiolazy import Stream

Endless iterables with operators (be careful with loops through an endless
iterator!):

.. code-block:: python

  >>> a = Stream(2, -2, -1) # Periodic
  >>> b = Stream(3, 7, 5, 4) # Periodic
  >>> c = a + b # Elementwise sum, periodic
  >>> c.take(15) # First 15 elements from the Stream object
  [5, 5, 4, 6, 1, 6, 7, 2, 2, 9, 3, 3, 5, 5, 4]

And also finite iterators (you can think on any Stream as a generator with
elementwise operators):

.. code-block:: python

  >>> a = Stream([1, 2, 3, 2, 1]) # Finite, since it's a cast from an iterable
  >>> b = Stream(3, 7, 5, 4) # Periodic
  >>> c = a + b # Elementwise sum, finite
  >>> list(c)
  [4, 9, 8, 6, 4]

LTI Filtering from system equations (Z-transform). After this, try summing,
composing, multiplying ZFilter objects:

.. code-block:: python

  >>> filt = 1 - z ** -1 # Diff between a sample and the previous one
  >>> filt
  1 - z^-1
  >>> data = filt([.1, .2, .4, .3, .2, -.1, -.3, -.2]) # Past memory has 0.0
  >>> data # This should have internally [.1, .1, .2, -.1, -.1, -.3, -.2, .1]
  <audiolazy.lazy_stream.Stream object at ...>
  >>> data *= 10 # Elementwise gain
  >>> [int(round(x)) for x in data] # Streams are iterables
  [1, 1, 2, -1, -1, -3, -2, 1]
  >>> data_int = filt([1, 2, 4, 3, 2, -1, -3, -2], zero=0) # Now zero is int
  >>> list(data_int)
  [1, 1, 2, -1, -1, -3, -2, 1]

LTI Filter frequency response plot (needs MatPlotLib):

.. code-block:: python

  (1 + z ** -2).plot().show()

.. image:: images/filt_plot.png

The ``matplotlib.figure.Figure.show`` method won't work unless you're
using a newer version of MatPlotLib (works on MatPlotLib 1.2.0), but you still
can save the above plot directly to a PDF, PNG, etc. with older versions
(e.g. MatPlotLib 1.0.1):

.. code-block:: python

  (1 + z ** -2).plot().savefig("my_plot.pdf")

On the other hand, you can always show the figure using MatPlotLib directly:

.. code-block:: python

  from matplotlib import pyplot as plt # Or "import pylab as plt"
  filt = 1 + z ** -2
  fig1 = filt.plot(plt.figure()) # Argument not needed on the first figure
  fig2 = filt.zplot(plt.figure()) # The argument ensures a new figure
  plt.show()

CascadeFilter instances and ParallelFilter instances are lists of filters with
the same operator behavior as a list, and also works for plotting linear
filters. Constructors accepts both a filter and an iterable with filters.
For example, a zeros and poles plot (needs MatPlotLib):

.. code-block:: python

  filt1 = CascadeFilter(0.2 - z ** -3) # 3 zeros
  filt2 = CascadeFilter(1 / (1 -.8 * z ** -1 + .6 * z ** -2)) # 2 poles
  # Here __add__ concatenates and __mul__ by an integer make reference copies
  filt = (filt1 * 5 + filt2 * 10) # 15 zeros and 20 poles
  filt.zplot().show()

.. image:: images/cascade_plot.png

Linear Predictive Coding (LPC) autocorrelation method analysis filter
frequency response plot (needs MatPlotLib):

.. code-block:: python

  lpc([1, -2, 3, -4, -3, 2, -3, 2, 1], order=3).plot().show()

.. image:: images/lpc_plot.png

Linear Predictive Coding covariance method analysis and synthesis filter,
followed by the frequency response plot together with block data DFT
(MatPlotLib):

.. code-block:: python

  >>> data = Stream(-1., 0., 1., 0.) # Periodic
  >>> blk = data.take(200)
  >>> analysis_filt = lpc.covar(blk, 4)
  >>> analysis_filt
  1 + 0.5 * z^-2 - 0.5 * z^-4
  >>> residual = list(analysis_filt(blk))
  >>> residual[:10]
  [-1.0, 0.0, 0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
  >>> synth_filt = 1 / analysis_filt
  >>> synth_filt(residual).take(10)
  [-1.0, 0.0, 1.0, 0.0, -1.0, 0.0, 1.0, 0.0, -1.0, 0.0]
  >>> amplified_blk = list(Stream(blk) * -200) # For alignment w/ DFT
  >>> synth_filt.plot(blk=amplified_blk).show()

.. image:: images/dft_lpc_plot.png

AudioLazy doesn't need any audio card to process audio, but needs PyAudio to
play some sound:

.. code-block:: python

  rate = 44100 # Sampling rate, in samples/second
  s, Hz = sHz(rate) # Seconds and hertz
  ms = 1e-3 * s
  note1 = karplus_strong(440 * Hz) # Pluck "digitar" synth
  note2 = zeros(300 * ms).append(karplus_strong(880 * Hz))
  notes = (note1 + note2) * .5
  sound = notes.take(int(2 * s)) # 2 seconds of a Karplus-Strong note
  with AudioIO(True) as player: # True means "wait for all sounds to stop"
    player.play(sound, rate=rate)

See also the docstrings and the "examples" directory at the github repository
for more help. Also, the huge test suite might help you understanding how the
package works and how to use it.

Project files
-------------

There are several files and directories in the AudioLazy repository (as well
as in the source distribution):

================ =============================================================
File/Directory   Description
================ =============================================================
audiolazy/       AudioLazy package modules source code
audiolazy/tests/ Testing subpackage
docs/            Documentation generation scripts
examples/        Examples for some AudioLazy use cases
images/          Images referenced by some reStructuredText documentation file
math/            Proof for some formula used by AudioLazy using Sympy CAS
CHANGES.rst      AudioLazy History, a.k.a. change log
conftest.py      Configuration for py.test, to work properly with doctests on
                 StrategyDict strategies and on an environment missing Numpy
COPYING.txt      License file
MANIFEST.in      List of extra distributed files to be included in the tarball
                 that doesn't need to be installed together with the package
README.rst       Some general information about the AudioLazy project
setup.py         General Python setup script for installation, testing, etc.
tox.ini          Configuration for tox, py.test and pytest-cov
.travis.yml      Travis-CI configuration (not in PyPI tarball/"egg" source
                 distribution)
================ =============================================================

The ``examples`` and the ``math`` directories might be useful for an AudioLazy
user. All Python files in these two directories are scripts intended to run on
both Python 2 and 3 unless they need something not yet available for Python 3
(e.g. wxPython), most of them have some external dependency.

----

Copyright (C) 2012-2016 Danilo de Jesus da Silva Bellini

License is GPLv3. See COPYING.txt for more details.


CHANGES.rst
..
  This file is part of AudioLazy, the signal processing Python package.
  Copyright (C) 2012-2016 Danilo de Jesus da Silva Bellini

  AudioLazy is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, version 3 of the License.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.

AudioLazy changes history
-------------------------

*** Unreleased ***

+ examples:

  - Make "modulo counter FM" example compatible with wxPython 3+

+ general:

  - Remove Travis CI configuration as it's no longer available
    for open source projects
  - Update tox for testing with CPython 2.7, CPython 3.2~3.10, PyPy and PyPy 3

*** Version 0.6 (Examples, STFT, phon2dB, tox, CI, wave file) ***

+ examples:

  - Formant synthesis for voiced "ah-eh-ee-oh-oo"
  - Musical keyboard synth example with a QWERTY keyboard (also via jack!)
  - Random synthesis with saving and memoization
  - Aesthetics for the Tkinter GUI examples
  - Matplotlib animated plot with mic input data (also works via jack!)
  - Perceptual experiment with butterworth filters (Scipy) applied to white
    noise (based on the Demonstrations to Auditory Scene Analysis)
  - Psychoacoustics using ISO/FDIS 226:2003

    * Equal loudness contour curves plotting
    * Constant phon chirp playing

  - Partial recreation of the "Windows and Figures of Merit" F. Harris
    comparison table and window plots
  - Playing/synthesizing "Ode to Joy" from its "score" written as code
  - All recording/playback examples now prepared for using both the default
    API and receiving it as an argument like "jack" or "alsa"
  - Realtime STFT (Short Time Fourier Transform) effects:

    * Robotize
    * "Roll" the magnitude spectrum while keeping the phase

+ general:

  - Tox for testing with CPython 2.7, CPython 3.2~3.6 and PyPy
  - Continuous integration with Travis CI and Coveralls
  - New ``_internals.py`` module to avoid exposing package internals together
    with the API, which also allowed the new simplified ``__init__.py``
  - Renewed setup.py in both its internals (e.g. using AST instead of
    string manipulation to avoid importing the package before installation)
    and its interface (e.g. the ``test`` command now calls ``tox``)
  - New ``conftest.py`` for testing doctests from strategies inside
    StrategyDict instances without the need of a ``__test__`` in the module
    and for skipping doctests known to need Numpy when it's not available
  - New ``math`` directory for adding scripts with symbolic math calculations
    (with Sympy) done for designing (or as a proof) for parts of the
    AudioLazy code. All lowpass and highpass strategies have their design
    explicitly explained there

+ lazy_analysis:

  - New ``overlap_add`` StrategyDict instance, allowing resynth after block
    processing/analysis, featuring block size auto-detection and window
    normalization to keep the output in the [-1; 1] range. Has 2
    implementations keeping the same interface:

    * ``numpy`` (*default*): needs Numpy arrays internally
    * ``list``: uses lists instead, doesn't need Numpy and was tested on Pypy

  - New ``stft`` StrategyDict instance, allowing Short Time Fourier Transform
    block processors / phase vocoder by:

    * Passing a function as the first parameter
    * Using a ``stft`` strategies as a decorator
    * Creating new strategies by avoiding the only positional parameter

    It was created with three Numpy-based strategies:

    * ``rfft``, ``real`` or ``base`` (*default*): needs ``numpy.fft.rfft``
      internally, as well as its inverse, to process a block in the
      frequency domain with values up to the Nyquist frequency, by assuming
      the input signal is real
    * ``cfft`` or ``complex``: Alike to the default approach but uses the
      ``numpy.fft.fft`` for a full complex-valued block in frequency domain,
      which means the output is a complex signal
    * ``cfftr`` or ``complex_real``: same to ``stft.cfft`` strategy, but
      keeps only the real part of the result

    Parameters allows changing the default zero-phase behavior (``before``
    and ``after`` functions), the transform and inverse transform functions,
    the overlap-add strategy (as well as its parameters), and obviously the
    block size and hop

  - The ``window`` StrategyDict now returns the "periodic" window data
    instead of the "symmetric" window to be used with the STFT directly
  - New ``wsymm`` StrategyDict with the same strategies from ``window`` but
    returning a "symmetric" window
  - Default ``window`` strategy is the Hann window (the same for ``wsymm``)
  - New ``cos`` strategy for ``window`` and ``wsymm`` implementing cosine to
    the power of alpha

+ lazy_auditory:

  - New ``phon2dB`` StrategyDict instance with the ISO/FDIS 226:2003 model
    for getting a SPL (Sound Pressure Level) in dBs from a phon value

+ lazy_core:

  - Add support to the new ``@`` matrix multiplication operator
  - ``OpMethod.get()`` now accepts numbers ``"1"`` and ``"2"`` as strings for
    unary and binary query for operators, and ``"all"`` is the new default, so
    ``OpMethod.get(without="2 ~")`` would get all operators but the [reversed
    or not] binary ones and the invert operator, which means it would yield
    only the unary ``"+"`` (pos) and ``-`` (neg), as probably was expected;
    OpMethod also had a bug fix regarding the shift operators

examples/fmbench.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# This file is part of AudioLazy, the signal processing Python package.
# Copyright (C) 2012-2016 Danilo de Jesus da Silva Bellini
#
# AudioLazy is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
"""
FM synthesis benchmarking
"""

from __future__ import unicode_literals, print_function
from timeit import timeit
import sys


# ===================
# Some initialization
# ===================
num_tests = 30
is_pypy = any(name.startswith("pypy") for name in dir(sys))
if is_pypy:
  print("PyPy detected!")
  print()
  numpy_name = "numpypy"
else:
  numpy_name = "numpy"


# ======================
# AudioLazy benchmarking
# ======================
kws = {}
kws["setup"] = """
from audiolazy import sHz, adsr, sinusoid
from math import pi
"""
kws["number"] = num_tests
kws["stmt"] = """
s, Hz = sHz(44100)
ms = 1e-3 * s
env = adsr(dur=5*s, a=20*ms, d=30*ms, s=.8, r=50*ms)
sin_data = sinusoid(freq=440*Hz,
                    phase=sinusoid(220*Hz) * pi)
result = sum(env * sin_data)
"""

print("=== AudioLazy benchmarking ===")
print("Trials:", kws["number"])
print()
print("Setup code:")
print(kws["setup"])
print()
print("Benchmark code (also executed once as 'setup'/'training'):")
kws["setup"] += kws["stmt"] # Helpful for PyPy
print(kws["stmt"])
print()
print("Mean time (milliseconds):")
print(timeit(**kws) * 1e3 / num_tests)
print("==============================")
print()


# ==================
# Numpy benchmarking
# ==================
kws_np = {}
kws_np["setup"] = "import {0} as np".format(numpy_name)
kws_np["number"] = num_tests
kws_np["stmt"] = """
rate = 44100
dur = 5 * rate
sustain_level = .8
# The np.linspace isn't in numpypy yet; it uses float64
attack = np.linspace(0., 1., num=int(np.round(20e-3 * rate)), endpoint=False)
decay = np.linspace(1., sustain_level, num=int(np.round(30e-3 * rate)),
                    endpoint=False)
release = np.linspace(sustain_level, 0., num=int(np.round(50e-3 * rate)),
                      endpoint=False)
sustain_dur = dur - len(attack) - len(decay) - len(release)
sustain = sustain_level * np.ones(sustain_dur)
env = np.hstack([attack, decay, sustain, release])
freq220 = 220 * 2 * np.pi / rate
freq440 = 440 * 2 * np.pi / rate
phase220 = np.arange(dur, dtype=np.float64) * freq220
phase440 = np.arange(dur, dtype=np.float64) * freq440
sin_data = np.sin(phase440 + np.sin(phase220) * np.pi)
result = np.sum(env * sin_data)
"""

# Alternative for numpypy (since it don't have "linspace" nor "hstack")
stmt_npp = """
rate = 44100
dur = 5 * rate
sustain_level = .8
len_attack = int(round(20e-3 * rate))
attack = np.arange(len_attack, dtype=np.float64) / len_attack
len_decay = int(round(30e-3 * rate))
decay = (np.arange(len_decay - 1, -1, -1, dtype=np.float64
                  ) / len_decay) * (1 - sustain_level) + sustain_level
len_release = int(round(50e-3 * rate))
release = (np.arange(len_release - 1, -1, -1, dtype=np.float64
                    ) / len_release) * sustain_level
env = np.ndarray(dur, dtype=np.float64)
env[:len_attack] = attack
env[len_attack:len_attack+len_decay] = decay
env[len_attack+len_decay:dur-len_release] = sustain_level
env[dur-len_release:dur] = release
freq220 = 220 * 2 * np.pi / rate
freq440 = 440 * 2 * np.pi / rate
phase220 = np.arange(dur, dtype=np.float64) * freq220
phase440 = np.arange(dur, dtype=np.float64) * freq440
sin_data = np.sin(phase440 + np.sin(phase220) * np.pi)
result = np.sum(env * sin_data)
"""

try:
  if is_pypy:
    import numpypy as np
  else:
    import numpy as np
except ImportError:
  print("Numpy not found. Finished benchmarking!")
else:
  if is_pypy:
    kws_np["stmt"] = stmt_npp

  print("Numpy found!")
  print()
  print("=== Numpy benchmarking ===")
  print("Trials:", kws_np["number"])
  print()
  print("Setup code:")
  print(kws_np["setup"])
  print()
  print("Benchmark code (also executed once as 'setup'/'training'):")
  kws_np["setup"] += kws_np["stmt"] # Helpful for PyPy
  print(kws_np["stmt"])
  print()
  print("Mean time (milliseconds):")
  print(timeit(**kws_np) * 1e3 / num_tests)
  print("==========================")
